File: algorithms/heap/minimum_number_of_refueling_stops.cc
// https://leetcode.com/problems/minimum-number-of-refueling-stops/
// https://leetcode.com/problems/minimum-number-of-refueling-stops/discuss/149839/DP-O(N2)-and-Priority-Queue-O(NlogN)

// time complexity: O(nlog(n))

#include <queue>
#include <vector>

using namespace std;

class Solution {
public:
    int minRefuelStops(int target, int startFuel, vector<vector<int>> &stations) {
        int fuel = startFuel, i = 0, len = stations.size(), n_refuel = 0;
        priority_queue<int> pq;
        for (; fuel < target; n_refuel++) {
            while (i < len && fuel >= stations[i][0])
                pq.push(stations[i++][1]);
            if (pq.empty())
                return -1;
            fuel += pq.top();
            pq.pop();
        }
        return n_refuel;
    }
};

File: algorithms/heap/minimum_number_of_refueling_stops.py
# https://leetcode.com/problems/minimum-number-of-refueling-stops/
# https://leetcode.com/problems/minimum-number-of-refueling-stops/discuss/149839/DP-O(N2)-and-Priority-Queue-O(NlogN)

# time complexity: O(nlog(n))

import heapq

class Solution:
    def minRefuelStops(self, target: int, startFuel: int, stations: list[list[int]]) -> int:
        n_refuel, i, fuel = 0, 0, startFuel
        pq = []
        while fuel < target:
            while i < len(stations) and fuel >= stations[i][0]:
                heapq.heappush(pq, -stations[i][1])
                i += 1
            if not pq:
                return -1
            fuel -= heapq.heappop(pq)
            n_refuel += 1
        return n_refuel